### ISSUE TITLE
Uncaught TypeError: Cannot read property 'length' of null (CollisionTile in Tile.reloadSymbolData)

### ISSUE THREAD

**Issue by sguignot** [mapbox/mapbox-gl-js#3700]
**mapbox-gl-js version**: 0.28
Not reproduced with 0.26 ; Maybe present in 0.27 too.

### Steps to Trigger The Error

 1. Setup a geojson layer with a lot of features: for example, every building polygons from OSM in a city like Paris, France.
 2. Zoom in and out randomly inside the city
 3. Sometimes the following error occurs (details and stack below): `Uncaught TypeError: Cannot read property 'length' of null`

### Error
```
Uncaught TypeError: Cannot read property 'length' of null
    at new CollisionTile
    at Tile.reloadSymbolData
    at Tile.done
    at Actor.receive
```

```
var CollisionTile = function CollisionTile(angle, pitch, collisionBoxArray) {
    if (typeof angle === 'object') {
        var serialized = angle;
        collisionBoxArray = pitch;
        angle = serialized.angle;
        pitch = serialized.pitch;
        this.grid = new Grid(serialized.grid);
        this.ignoredGrid = new Grid(serialized.ignoredGrid);
    } else {
        this.grid = new Grid(EXTENT, 12, 6);
        this.ignoredGrid = new Grid(EXTENT, 12, 0);
    }

    this.minScale = 0.5;
    this.maxScale = 2;

    this.angle = angle;
    this.pitch = pitch;

    var sin = Math.sin(angle),
        cos = Math.cos(angle);
    this.rotationMatrix = [cos, -sin, sin, cos];
    this.reverseRotationMatrix = [cos, sin, -sin, cos];

    // Stretch boxes in y direction to account for the map tilt.
    this.yStretch = 1 / Math.cos(pitch / 180 * Math.PI);

    // The amount the map is squished depends on the y position.
    // Sort of account for this by making all boxes a bit bigger.
    this.yStretch = Math.pow(this.yStretch, 1.3);

    this.collisionBoxArray = collisionBoxArray;
    if (collisionBoxArray.length === 0) { // <===== !!!! CRASH HERE !!!!
        // the first collisionBoxArray is passed to a CollisionTile
```

<img width="1351" alt="stack_0" src="https://cloud.githubusercontent.com/assets/986704/20628229/0e0598ca-b325-11e6-924b-41de764aadd5.png">

<img width="1350" alt="stack_1" src="https://cloud.githubusercontent.com/assets/986704/20628242/1af11f00-b325-11e6-8741-c34c1a666d3b.png">

<img width="1351" alt="stack_2" src="https://cloud.githubusercontent.com/assets/986704/20628247/21b1cb6e-b325-11e6-84fc-ca284af90af4.png">

**Comment by AndyMoreland** [comment]
I see this as well.

**Comment by mollymerp** [comment]
@sguignot thanks for reporting this issue! could you please provide a jsfiddle / jsbin that reproduces this error? 

it is possible that https://github.com/mapbox/mapbox-gl-js/pull/3681 fixes this exception, so I'd like to verify before investigating further. 

thanks!

**Comment by lbud** [comment]
👍 after some investigation, I suspect this is a different issue from #3681, but we haven't been able to reproduce it so far 😬 a jsfiddle doing so would be very helpful!

**Comment by lucaswoj** [comment]
Closing pending reproduction steps

**Comment by sguignot** [comment]
**mapbox-gl-js versions**: 0.27, 0.28, 0.29, 0.30
Not reproduced with 0.26, this bug is present in all versions since 0.27.

### Steps to Trigger The Error

 1. Get a mouse device that supports scrolling with high speed / velocity / inertia (I am using a Magic mouse 2 from Apple: http://www.apple.com/shop/product/MLA02LL/A/magic-mouse-2).
 2. Go to the following blog demo: https://www.mapbox.com/bites/00304/
 3. Open the browser console and add "length" as error filter to avoid the flood of 404 errors.
 4. Quickly move the camera where features are located: zoom in and out quickly, pan and change the camera pitch. In less than a minute of zooming and moving the camera I reproduce the following error in the console: `Uncaught TypeError: Cannot read property 'length' of null` (browsers on which I reproduce this error: Google Chrome 55 and Safari 10).

### Error
```
Uncaught TypeError: Cannot read property 'length' of null
    at new CollisionTile
    at Tile.reloadSymbolData
    at Tile.done
    at Actor.receive
```

<img width="1440" alt="screen shot 2017-01-10 at 15 04 40" src="https://cloud.githubusercontent.com/assets/986704/21809918/0bbaa62a-d74a-11e6-9fc2-7cdeda0b3413.png">

**Comment by musicformellons** [comment]
I am getting this as well (version 0.30.0):
```
Uncaught TypeError: Cannot read property 'length' of null
    at new CollisionTile (eval at <anonymous> (0.50f3799….js:263), <anonymous>:157:670)
    at Tile.reloadSymbolData (eval at <anonymous> (0.50f3799….js:263), <anonymous>:101:1495)
    at Tile.t (eval at <anonymous> (0.50f3799….js:263), <anonymous>:101:2104)
    at Actor.receive (eval at <anonymous> (0.50f3799….js:263), <anonymous>:213:738)
CollisionTile @ mapbox-gl.js?31ef:157
Tile.reloadSymbolData @ mapbox-gl.js?31ef:101
t @ mapbox-gl.js?31ef:101
Actor.receive @ mapbox-gl.js?31ef:213
```

**Comment by mourner** [comment]
@sguignot can you please try reproducing this again and posting the properties of `this` (`Tile` instance) on the breakpoint where it happens (in either `done` or `reloadSymbolData`)? I'm specifically interested in the values of `this.state` and `this.collisionBoxArray`.

**Comment by sguignot** [comment]
@mourner sorry for the delay.
I ran the demo "Restaurants Complaints NYC" with https://api.tiles.mapbox.com/mapbox-gl-js/v0.30.0/mapbox-gl-dev.js
I put conditional breakpoint on line 49 of file collision_tile.js (condition is `!collisionBoxArray`)
`line 49: if (collisionBoxArray.length === 0) {`

When the error occurs (ie `!collisionBoxArray`), I have `this.state : "reloading"` and `this. collisionBoxArray : null`. You will find the detailed context for each frame of the call stack below:

<img width="1440" alt="mapbox_stack0_collitiontile_ctor" src="https://cloud.githubusercontent.com/assets/986704/21989655/0a29b096-dc0c-11e6-8be9-b2a65e83797b.png">
<img width="1440" alt="mapbox_stack1_reloadsymboldata" src="https://cloud.githubusercontent.com/assets/986704/21989660/0c851f24-dc0c-11e6-9418-2a68b334a338.png">
<img width="1440" alt="mapbox_stack2_done" src="https://cloud.githubusercontent.com/assets/986704/21989662/0ee98516-dc0c-11e6-9734-7cc8ad141352.png">

**Comment by sguignot** [comment]
@mourner it seems a lot easier to reproduce this error when spinning the map using the right click, check out the following screencast:
![mapbox-gl-js_issue_3700_howto_reproduce_short](https://cloud.githubusercontent.com/assets/986704/21992363/6df45248-dc16-11e6-8883-a741c04565e1.gif)

**Comment by mourner** [comment]
@sguignot finally managed to reproduce, thanks a lot for the help! On it now.

**Comment by GuillaumeTS** [comment]
Reproduced with the same steps with mapbox 0.37 and 0.38.

**Comment by lbud** [comment]
I can also reproduce on master.

**Comment by lbud** [comment]
https://jsbin.com/sujonoyupa/edit?html,output — error appears after clicking the button a few times.

**Comment by sguignot** [comment]
@lbud Thanks for the jsbin, do you have any news regarding the bugfix?

We have a lot of logs/alerts in production because of this regression, it would be great to have a fix in the next version (v0.39).

**Comment by edap** [comment]
I'm having the same problem in production

**Comment by mollymerp** [comment]
@sguignot @edap would you be willing to see if the error is still happening on master? I think it was fixed in https://github.com/mapbox/mapbox-gl-js/commit/caf4cfb058b5060dffee5ccb7795a9b5add7d20a as I can no longer reproduce on master.

**Comment by mollymerp** [comment]
Closing for now – please reopen if you continue to see this issue after the next release (this afternoon)

**Comment by brncsk** [comment]
This is still happening for me on `0.39.1`.

**Comment by sguignot** [comment]
Thanks @mollymerp it's fixed on my side, I just tested with 0.39.1 and errors are gone.

For quick testing, you can reproduce the error in 0.38 here after a few clicks:
https://jsbin.com/sonuvoyofe/1/edit?html,output

Here in 0.39.1 you can check error is gone:
https://jsbin.com/hoyaputude/1/edit?html,output

I'm going to deploy 0.39.1 in production soon, then after a few days I will confirm if the error is completely gone or if it persists (there may be several root causes after all).

**Comment by EbilPanda** [comment]
yeah also happening to me occasional when zooming or using the fly-method...
I'm using custom tileset and style.

Error:
> ERROR TypeError: Cannot read property 'length' of null
    at new CollisionTile (mapbox-gl.js:337)
    at Function.CollisionTile.deserialize (mapbox-gl.js:337)
    at Tile.reloadSymbolData (mapbox-gl.js:215)
    at mapbox-gl.js:215
    at Actor.receive (mapbox-gl.js:397)
    at t.invokeTask (polyfills.js:3)
    at Object.onInvokeTask (core.es5.js:4140)
    at t.invokeTask (polyfills.js:3)
    at r.runTask (polyfills.js:3)
    at Worker.invoke (polyfills.js:3)

using 0.39.1 via npm, with webpack 3.4.1 and excluded mapbox from uglifier (`exports.module.noParse = /(mapbox-gl)\.js$/;`) in an ionic2-app (angular4)

**Comment by mollymerp** [comment]
@EbilPanda @brncsk sorry you're still running into this – can you cut a new issue with a jsfiddle/jsbin that can reproduce the issue?

**Comment by EbilPanda** [comment]
i'm sorry I'm currently not able to reproduce the error on fiddl/bin :(
I think the problem is with webpack/uglify that stills running over the already uglified code and then occures error, I think  I will go with the cdn to test if it happends there too

**Comment by EbilPanda** [comment]
But i found the error line in the mapbox-gl-devjs: Line 23052 `if (collisionBoxArray.length === 0)`
I've attached to scope-variables directly before the error occured (Line 23051).
[mapbox.scope.json.txt](https://github.com/mapbox/mapbox-gl-js/files/1182953/mapbox.scope.json.txt)

**Comment by shf123** [comment]
I also still happens for me in version 0.39.1.

I made an example in jsfiddle:
https://jsfiddle.net/stighef/y5z4cwLt/

A FeatureCollection with 10 000 rectangles are toggled on and off when
the zoomend-event is triggered. 

After zooming out and in, I get lots of these:

![image](https://user-images.githubusercontent.com/1443522/28723651-ac073c56-73b7-11e7-8224-bc40c96940ef.png)

**Comment by sguignot** [comment]
@mollymerp I reopen the issue because the jsfiddle from @shf123 triggers the error with version 0.39.1

**Comment by snkashis** [comment]
Yeah, still an issue for me in 0.39.1 as well.

**Comment by apolishch** [comment]
Yeah happening for me on 0.39.1 as well

**Comment by mourner** [comment]
Thanks a lot for the new test case — got it to immediately reproduce! Will make fixing much easier.

**Comment by mourner** [comment]
Hey everyone, we have a fix in #5185 — until we ship this in a release, feel free to use this patch for a custom build of GL JS if you experience this problem.
